File: kts_provider/src/main/java/com/lasley/kts_provider/data/Artist.java
package com.lasley.kts_provider.data;


import androidx.annotation.NonNull;
import androidx.room.ColumnInfo;
import androidx.room.Entity;
import androidx.room.PrimaryKey;
import androidx.room.Relation;

import com.google.gson.annotations.SerializedName;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.UUID;

@Entity(tableName = "_artist")
public class Artist {
    @PrimaryKey
    @ColumnInfo(name = "artist_id")
    @SerializedName("artist_id")
    @NonNull
    public String uuid = UUID.randomUUID().toString();

    public String name;
    @ColumnInfo(name = "album_ids", index = true)
    @SerializedName("album_ids")
    public String albums = "";

    public long createdTime = System.currentTimeMillis();
    public long updatedTime = System.currentTimeMillis();

    public Artist() {
    }

    public Artist(String name) {
        this.name = name;
    }

    public Artist(String name, List<Album> albums) {
        this(name);
//        this.albums.addAll(albums);
    }

    public List<String> albumIDs() {
        if (albums.isEmpty()) return new ArrayList<>();
        return Arrays.asList(albums.split(","));
    }

    public boolean removeAlbum(String id) {
        ArrayList<String> ids = new ArrayList<>(albumIDs());
        if (ids.contains(id) && ids.remove(id)) {
            albums = idsToString(ids);
            return true;
        } else
            return false;
    }


    public boolean addAlbum(String id) {
        ArrayList<String> ids = new ArrayList<>(albumIDs());
        if (ids.contains(id)) return false;
        ids.add(id);
        albums = idsToString(ids);
        return true;
    }

    private String idsToString(List<String> ids) {
        StringBuilder newAlbums = new StringBuilder();
        for (String item : ids) {
            newAlbums.append(item).append(",");
        }
        // remove final ","
        if (newAlbums.length() > 1)
            newAlbums.deleteCharAt(newAlbums.length() - 1);
        return newAlbums.toString();
    }
}

File: kts_viewer/src/main/kotlin/com/lasley/kts_viewer/data/Artist.kt
package com.lasley.kts_viewer.data

import com.google.gson.annotations.SerializedName
import java.util.UUID

data class Artist(
    var name: String = ""
) : CommonInf() {

    @SerializedName("artist_id")
    override var uuid: String = UUID.randomUUID().toString()

    @SerializedName("album_ids")
    var albumIDs = ""
        private set

    val albums: List<Album>
        get() = albumID_List.mapNotNull { LocalDatabase[it] as? Album }

    val albumID_List: List<String>
        get() = albumIDs.split(",")

    fun addAlbum(vararg album: Album) {
        albumIDs = (albumID_List + album.map { it.uuid }).distinct().joinToString(",")
    }

    /**
     * Removes a single instance of the specified element from this collection, if it is present.
     *
     * @return
     * - true: if the element has been successfully removed
     * - false: if it was not present in the collection
     */
    fun removeAlbum(album: Album): Boolean {
        val list = albumID_List.toMutableList()
        return if (list.remove(album.uuid)) {
            albumIDs = list.joinToString(",")
            true
        } else false
    }

    companion object {
        fun create(config: Artist.() -> Unit = {}): Artist {
            return Artist().apply(config).apply {
                uuid = UUID.randomUUID().toString()
            }
        }
    }
}
